README.md
#hapi-method-routes is a plugin that exposes server methods as routes

## Example:
```js
server.method('add', (a, b, done) => done(null, a.toString() + b.toString()), {});
server.inject({
  url: '/methods/',
  method: 'POST',
  payload: {
    method: 'add',
    values: [20, 25]
  }
}, response => {
  console.log(response.result.successful);
  console.log(response.result.result);
});
```

   will print 'true' and '2025'.  See tests for more examples.


CHANGELOG.md

0.0.3 / 2015-11-16
==================

  * hapi wraps cached server methods so that we can't get their argument length because of this we are dropping the parameter-length checking when calling a route-method.

0.0.2 / 2015-10-29
==================

  * allow passing in logs


index.js
'use strict';
var _ = require('lodash');
//
var defaults = {
  // option to change the endpoint:
  endpoint : '/methods',
  auth: null,
  // optional whitelist function to screen requests to a method:
  whitelist : function(methodName, request){
    // by default we always allow access:
    return true;
  },
  //  you can pass it a validator function
  validator : function(params){
    return params;
  }
}

exports.register = function(server, options, next) {
  // setup defaults:
  var settings = _.clone(options);
  settings = _.defaults(settings, defaults);
  var endpoint = settings.endpoint;
  var whitelist = settings.whitelist;
  var validator = settings.validator;
  // list the methods we're exporting:
  server.log(['hapi-method-routes', 'info'], "methodsRoutePlugin is exporting the following methods at :" + endpoint);

  // will decode any parameter list:
  function decodeParamList(paramList){
    return _.map(paramList, function(param){
      return decodeURIComponent(param);
    });
  }
  // call this to extra params from request by GET or POST:
  // note: I think we're mainly wanting to support JSON/POST requests
  function extractParamsFromRequest(request){
    if (request.method.toLowerCase()=='post'){
      return decodeParamList(request.payload.values);
    }
    return (!request.params.params) ? [] : decodeParamList(request.params.params.split("/"))
  }
  server.route({
      method: '*',
      path: endpoint + '/{methodName}/{params*}',
      config: {
        auth: settings.auth
      },
      handler: function (request, reply) {
        // get the method they are trying to call:
        var methodName = decodeURIComponent(request.params.methodName);
        // this goes at the top because we want to short-circuit if they don't have access
        // to the indicated method:
        if (!whitelist(methodName, request)){
          reply({successful: false, result: 'You do not have access to ' + methodName}).code(550);
          return;
        }
        // extract and validate any params:
        var params = extractParamsFromRequest(request);
        params = validator(params);
        var method = _.get(server.methods, methodName);

        // first we're going to check for obvious errors:
        if (params==undefined){
          reply({successful: false, result:'Unable to validate parameters for method ' + methodName}).code(500);
          return;
        }
        if (!method){
          reply({successful: false, result:'Method name ' + methodName + ' does not exist '}).code(404);
          return;
        }
        // add the 'done' callback to the function params
        // this is the method that will be called at the end of the method execution:
        params.push(function done(err,result){
           if (err){
             reply({successful: false, result:'Method name ' + methodName + ' threw this error: '  + err}).code(500);
           }
           else{
             reply({successful: true, result: result});
           }
        });
        // finally we try to execute the method:
        try{
          method.apply(null, params);
        }catch(exc){
          server.log(['hapi-method-routes', 'error'],exc);
          reply('Method name ' + methodName + ' failed. Error: ' + exc).code(404);
          return;
        }
      }
  });
  next();
}
exports.register.attributes = {
    pkg: require('./package.json')
};


example/example.js
"use strict"
var _ = require("lodash");
var Hapi = require("hapi");
var routeModule = require("../index.js");
var request = require("request");
var fs = require("fs");

// an example of a method to export. the last param of your method will be a 'done' callback
// who's first param is any error message and second is the output of the method
function add(a, b, done){
  var result = a+b;
  done(null, result);
}

// another example, this one will be added to server.method.math
// to demonstrate how methods can be namespaced:
function mathAdd(a,b,done){
  try{
    var floatA = parseFloat(a);
    var floatB = parseFloat(b);
    var result = floatA + floatB;
    done(null,result);
  }catch(exc){
    done(exc);
  }
}

var samplePayload = { "values" : [20,25]};

var server = new Hapi.Server({
  debug : {
       log : ['info', 'error']
     },
});

// here is how you register your methods for export:
server.method('add', add, {
  cache: {
    // cache : 'temp',
    expiresIn: 5000,
    generateTimeout : 5000
  }
});

server.method('math.add', mathAdd);
server.connection({ port: 3000 });
server.register(
  {
    register : routeModule,
    options : {
    }
  },
  function(err){
    server.log(['hapi-method-routes', 'info'], "module registered")
    if (err) {
      server.log(['hapi-method-routes', 'error'],err);
    }
    else
      server.start(function () {
        console.log('Server running at: %s, beginning tests...', server.info.uri);
        // test the method:
        request({
            url: 'http://localhost:3000/methods/add', //URL to hit
            method: 'POST',
            //Lets post the following key/values as form
            json: samplePayload
        }, function(error, response, body){
            if(error)
                console.log(error);
            else if (body.successful)
              console.log("add, result should be '2025': %s", body.result);
            else
              console.log("add did not succeed, result was : %s", response.statusCode, body.result);
        });
        /*
        // test namespace method:
        request({
            url: 'http://localhost:3000/methods/math.add', //URL to hit
            method: 'POST',
            //Lets post the following key/values as form
            json: samplePayload
        }, function(error, response, body){
            if(error)
              console.log(error);
            else if (body.successful)
              console.log("math.add result should be 45:  %s", body.result);
            else
              console.log("math.add did not succeed, result was : %s %s", response.statusCode, body.result);
        });
        // test calling a method that doesn't exist:
        request({
            url: 'http://localhost:3000/methods/math.doesntexist', //URL to hit
            method: 'POST',
            json: samplePayload,
        }, function(error, response, body){
            if(error) {
              console.log(error);
            } else {
              console.log("Unavailable method should return 404: doesn't exist: ", response.statusCode, body.result);
            }
        });
        // test calling a method with no parameters
        request({
            url: 'http://localhost:3000/methods/math.add', //URL to hit
            method: 'POST',
            json: {},
        }, function(error, response, body){
            if(error)
              console.log(error);
            else
              console.log("empty json should throw an error: ", response.statusCode, body.result);
        });
*/
    });
  }
);
